tests/test_cache.py

```diff
@@ -158,3 +150,10 @@ def test_save_doesnt_exists(self, mock_dumps, mock_dump, mock_exists, mock_mkdir
         self.assertEqual(mock_mkdir.call_count, 1)
         self.assertEqual(mock_dump.call_count, 1)
         self.assertEqual(mock_dumps.call_count, 1)
+
+    def temporary_file_with_cleanup(self, **kwargs):
+        t = tempfile.NamedTemporaryFile(delete=False, **kwargs)
```

Comment:
could you comment on why the delete=False is needed, afaict the only one of these that would need this is test_get_set because it opens the named temp file again, but none of them is doing a read after close operation that i can see.

Reply:
**tldr;** For a few of the cache tests there was an option to remove Python 2.7 support and make them pass on Windows with `delete=True`, but it added relatively complex code and was very fragile (if anyone added a file read into the underlying cache library or extended the test it would no longer function without refactoring).  So I opted to just be consistent with `delete=False` everywhere, especially since we already had the simple/reliable `addcleanup` in use.

**Details:**

Many of those cache tests had this
```
temp_cache_file = tempfile.NamedTemporaryFile()
self.addCleanup(temp_cache_file.close)
```
And then they did a single additional `open` which failed with a `Permission Denied` on Windows.

So when a `NamedTemporaryFile` with `delete=True` is created on Windows it is opened with the flag `FILE_FLAG_DELETE_ON_CLOSE`.   Per Win32 documentation on that flag:

> Subsequent open requests for the file fail, unless the FILE_SHARE_DELETE share mode is specified.

So even for a single additional write to the temporary file (`cache.save`) you have to pass this special flag to succeed, which requires you to update all your `open` calls with a custom `opener`.  

At first I took this route (I think you had tried it somewhere in GCP code also), but you run into two problems:

A) This custom `opener` override isn't in `builtins.open` on Python 2.7 and the workaround is really *really* bad.
B) Passing this flag only works once, because if you close that new handle the file is instantly deleted and the next attempt at `open` fails.

So basically even if you give up Python 2 support (which I was fine with) - there is **no** way to get the behavior of "delete file after all handles are closed".

So for these cache tests it made more sense to have a second cleanup to `unlink` than to make the code much more complicated and lose Py27.  Also, for a few of the cache tests there was no option due to 3 file handles being opened.

Reply:
Thanks for the exhaustive explanation, yeah with those details in mind and looking at the cache implementation, this seems like the most expedient road. The cache class impl would need to change significantly to holding open a file handle to be compatible, i hadn't realized its doing a open/write/flush every save call.